src/ruby/puppetserver-lib/puppet/server/master.rb

```diff
@@ -66,13 +73,19 @@ def handleRequest(request)
   end
 
   def compileCatalog(request_data)
-    @catalog_compiler.compile(convert_java_args_to_ruby(request_data))
+    Puppet.override(lookup_key_recorder: create_recorder) do
+      @catalog_compiler.compile(convert_java_args_to_ruby(request_data))
+    end
   end
 
   def compileAST(compile_options)
     Puppet::Server::ASTCompiler.compile(convert_java_args_to_ruby(compile_options))
   end
 
+  def create_recorder
+    @track_lookups ? Puppet::Server::KeyRecorder.new : Puppet::Pops::Lookup::KeyRecorder.singleton
```

Comment:
Why not `nil`? The Puppet code that calls this uses a default value right? Should we remove the default value in Puppet?